### PR TITLE
[FSDP] Fix summon_full_params test

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74452
* #74517
* __->__ #74456

This test did not actually do any CPU offloading. Adding it revealed
an issue in no_shard (currently only when world_size == 1) case which is
tracked for a fix in https://github.com/pytorch/pytorch/issues/74166

Differential Revision: [D35003793](https://our.internmc.facebook.com/intern/diff/D35003793/)

### CODE REVIEW

**Review by zhaojuanmao — APPROVED** [review]

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 1]
```diff
@@ -38,9 +38,11 @@
 
 
 def _run_test_summon_full_param_writeback(cls, writeback, cpu_offload, modify_outer):
-    model = FSDP(
-        nn.Sequential(FSDP(nn.Linear(5, 5, bias=False)), nn.Linear(5, 3, bias=False))
-    ).cuda(cls.rank)
+    lin1 = FSDP(nn.Linear(5, 5, bias=False).cuda(cls.rank), cpu_offload=cpu_offload)
+    lin2 = nn.Linear(5, 3, bias=False).cuda(cls.rank)
+    model = FSDP(nn.Sequential(lin1, lin2), cpu_offload=cpu_offload)
```
Comment (mrshenli) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:43]: curious, what's the expectation on the input model when `cpu_offload==True`? Does it have to reside on GPU? Or is it OK to pass in a CPU model?

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 2]
```diff
@@ -38,9 +38,11 @@
...
+    if not cpu_offload:
+        model = model.cuda(cls.rank)
```
Comment (mrshenli) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:45]: why do we need this? I assume when `cpu_offload==False`, the model is already on cuda, no?

File: test/distributed/fsdp/test_fsdp_summon_full_params.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:43]: Passing CPU offload with CPU model should work, as we don't assume any device in the constructor and would restore to GPU in forward pass before all_gather. Although, it does not seem that we have a test for this, we should add one.

File: test/distributed/fsdp/test_fsdp_summon_full_params.py (hunk 2 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:45]: Right, we can remove this because lin1, lin2 are already on CUDA device.

**Pull request closed without merging**